### PR TITLE
[SPARK-18120][SPARK-19557][SQL] Call QueryExecutionListener callback methods for DataFrameWriter methods

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

We only notify `QueryExecutionListener` for several `Dataset` operations, e.g. collect, take, etc. We should also do the notification for `DataFrameWriter` operations.

## How was this patch tested?

new regression test

close https://github.com/apache/spark/pull/16664

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @salilsurendran @vanzin @gatorsmile @sameeragarwal

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -211,13 +211,15 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
     assertNotBucketed("save")
-    val dataSource = DataSource(
-      df.sparkSession,
-      className = source,
-      partitionColumns = partitioningColumns.getOrElse(Nil),
-      options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    runCommand(df.sparkSession, "save") {
+      SaveIntoDataSourceCommand(
```
Comment (vanzin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Does this also cover SPARK-19557? If so, might as well mention that in the PR (or close the bug as a duplicate or related or something).

**Review by vanzin — COMMENTED** [review]
Looks good as far as I can tell. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -573,6 +575,24 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     format("csv").save(path)
   }
 
+  private def runCommand(session: SparkSession, name: String)(command: LogicalPlan): Unit = {
+    val qe = session.sessionState.executePlan(command)
+    try {
+      qe.executedPlan.foreach { plan =>
+        plan.resetMetrics()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: just realized that, in this code path, we pass in a logical plan and will always get a new physical plan, so we don't need to reset metrics here, let me remove it

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SaveIntoDataSourceCommand.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  override protected def innerChildren: Seq[QueryPlan[_]] = Seq(query)
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    DataSource(
+      sparkSession,
+      className = provider,
+      partitionColumns = partitionColumns,
+      options = options).write(mode, Dataset.ofRows(sparkSession, query))
+
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SaveIntoDataSourceCommand.scala]: Do we need to  Invalidate the cache to be consistent with `InsertIntoDataSourceCommand`?
```
    sparkSession.sharedState.cacheManager.invalidateCache(query)
```

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 4]
```diff
@@ -573,6 +575,21 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     format("csv").save(path)
   }
 
+  private def runCommand(session: SparkSession, name: String)(command: LogicalPlan): Unit = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Add a function description like?

  /**
   * Wrap a DataFrameWriter action to track the QueryExecution and time cost, then report to the
   * user-registered callback functions.
   */

File: sql/core/src/test/scala/org/apache/spark/sql/util/DataFrameCallbackSuite.scala [hunk 5]
```diff
@@ -159,4 +161,56 @@ class DataFrameCallbackSuite extends QueryTest with SharedSQLContext {
 
     spark.listenerManager.unregister(listener)
   }
+
+  test("execute callback functions for DataFrameWriter") {
+    val commands = ArrayBuffer.empty[(String, LogicalPlan)]
+    val exceptions = ArrayBuffer.empty[(String, Exception)]
+    val listener = new QueryExecutionListener {
+      // Only test successful case here, so no need to implement `onFailure`
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/util/DataFrameCallbackSuite.scala]: invalid comment?

**Comment by gatorsmile** [comment]
LGTM except three minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SaveIntoDataSourceCommand.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SaveIntoDataSourceCommand.scala]: we don't have the `LogicalRelation` to be used as cache key.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SaveIntoDataSourceCommand.scala]: : )

**Comment by gatorsmile** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 4 above)
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Why don't we use `SQLExecution` instead of this?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: the problem is that, in some commands, like `InsertIntoHiveTable`, we already use `SQLExecution.newExecution`, so we can't use it again to wrap these commands.

In the future we should figure out a central place to put `SQLExecution.newExecution`

**Pull request closed without merging**